// conan-js-core/dist/lib/conan-monitor/domain/monitorFacade.js
export class MonitorFacade {
    constructor(monitor, actions) {
        this.monitor = monitor;
        this.actions = actions;
    }
    start() {
        this.monitor.start();
        return this;
    }
    addAsyncReaction(def) {
        return this.monitor.addAsyncReaction(def);
    }
    addReaction(def) {
        return this.monitor.addReaction(def);
    }
    once(reaction, name) {
        this.monitor.once(reaction, name);
        return this;
    }
    get do() {
        return this.monitor.do;
    }
    getEvents() {
        return this.monitor.getEvents();
    }
    getName() {
        return this.monitor.getName();
    }
    stop(eventsConsumer) {
        this.monitor.stop(eventsConsumer);
    }
    getData() {
        return this.monitor.getData();
    }
    next(cb) {
        this.monitor.next(cb);
    }
    get asyncThread() {
        return this.monitor.asyncThread;
    }
    get mainThread() {
        return this.monitor.mainThread;
    }
    get metaFlow() {
        return this.monitor.metaFlow;
    }
    chain(operation, name) {
        return this.monitor.mainThread.chain(operation, name);
    }
    transaction(code) {
        return this.monitor.transaction(code);
    }
    openTransaction(name) {
        this.monitor.openTransaction(name);
    }
    closeTransaction(callback) {
        this.monitor.closeTransaction(callback);
    }
}
