// conan-js-core/dist/lib/conan-thread/domain/threadFacade.js
export class ThreadFacade {
    constructor(thread, actions) {
        this.thread = thread;
        this.actions = actions;
    }
    chain(mutatorsCb, name) {
        this.thread.chain(mutatorsCb, name);
        return undefined;
    }
    next(cb) {
        this.thread.next(cb);
    }
    start() {
        this.thread.start();
        return this;
    }
    stop(eventsConsumer) {
        this.thread.stop(eventsConsumer);
        return this;
    }
    get do() {
        let steps = this.reducers;
        return Object.assign(Object.assign({}, steps), this.actions);
    }
    get reducers() {
        return this.thread.reducers;
    }
    addReaction(def) {
        return this.thread.addReaction(def);
    }
    get isRunning() {
        return this.thread.isRunning;
    }
    getData() {
        return this.thread.getData();
    }
    getEvents() {
        return this.thread.getEvents();
    }
    getName() {
        return this.thread.getName();
    }
    changeLoggingNature(nature) {
        this.thread.changeLoggingNature(nature);
    }
    log(msg) {
        this.thread.log(msg);
    }
    once(reaction, name) {
        this.thread.once(reaction, name);
        return this;
    }
    createRuntimeTracker(runtimeEvent, payload) {
        return this.thread.createRuntimeTracker(runtimeEvent, payload);
    }
    monitor(toMonitor, thenCallback, name, payload) {
        return this.thread.monitor(toMonitor, thenCallback, name, payload);
    }
    addReactionNext(def) {
        this.thread.addReactionNext(def);
        return this;
    }
    getDefinition() {
        return this.thread.getDefinition();
    }
}
