// conan-js-core/dist/lib/conan-flow/domain/flowRuntimeEvents.js
export var FlowRuntimeEventSource;
(function (FlowRuntimeEventSource) {
    FlowRuntimeEventSource["FLOW_CONTROLLER"] = "FLOW_CONTROLLER";
    FlowRuntimeEventSource["FLOW_THREAD"] = "FLOW_THREAD";
    FlowRuntimeEventSource["CONTEXT"] = "CONTEXT";
    FlowRuntimeEventSource["FLOW_FACTORY"] = "FLOW_FACTORY";
})(FlowRuntimeEventSource || (FlowRuntimeEventSource = {}));
export var FlowRuntimeEventTiming;
(function (FlowRuntimeEventTiming) {
    FlowRuntimeEventTiming["REQUEST_START"] = "REQUEST_START";
    FlowRuntimeEventTiming["REQUEST_END"] = "REQUEST_END";
    FlowRuntimeEventTiming["REQUEST_CANCEL"] = "REQUEST_CANCEL";
    FlowRuntimeEventTiming["USER_TRACE"] = "USER_TRACE";
    FlowRuntimeEventTiming["TRACE"] = "TRACE";
    FlowRuntimeEventTiming["INFO"] = "INFO";
})(FlowRuntimeEventTiming || (FlowRuntimeEventTiming = {}));
export var FlowRuntimeEventType;
(function (FlowRuntimeEventType) {
    FlowRuntimeEventType["MONITOR"] = "MONITOR";
    FlowRuntimeEventType["PROCESS_REACTIONS"] = "PROCESS_REACTIONS";
    FlowRuntimeEventType["START"] = "START";
    FlowRuntimeEventType["REQUEST_STATUS"] = "REQUEST_STATUS";
    FlowRuntimeEventType["FLAG_AS_SETTLED"] = "FLAG_AS_SETTLED";
    FlowRuntimeEventType["USER_MSG"] = "USER_MSG";
    FlowRuntimeEventType["REQUEST_TRANSITION"] = "REQUEST_TRANSITION";
    FlowRuntimeEventType["CREATE_FLOW"] = "CREATE_FLOW";
})(FlowRuntimeEventType || (FlowRuntimeEventType = {}));
