// conan-js-core/dist/lib/conan-flow/domain/flowRuntimeEvents.js
export var FlowEventNature;
(function (FlowEventNature) {
    FlowEventNature["MAIN"] = "MAIN";
    FlowEventNature["META"] = "META";
    FlowEventNature["HELPER"] = "HELPER";
    FlowEventNature["ASYNC"] = "ASYNC";
    FlowEventNature["AUX"] = "AUX";
    FlowEventNature["ASAP"] = "ASAP";
})(FlowEventNature || (FlowEventNature = {}));
export var FlowEventSource;
(function (FlowEventSource) {
    FlowEventSource["FLOW_CONTROLLER"] = "FLOW_CONTROLLER";
    FlowEventSource["FLOW_THREAD"] = "FLOW_THREAD";
    FlowEventSource["CONTEXT"] = "CONTEXT";
    FlowEventSource["USER_MSG"] = "USER_MSG";
    FlowEventSource["FLOW_FACTORY"] = "FLOW_FACTORY";
})(FlowEventSource || (FlowEventSource = {}));
export var FlowEventLevel;
(function (FlowEventLevel) {
    FlowEventLevel["DEBUG"] = "DEBUG";
    FlowEventLevel["TRACE"] = "TRACE";
    FlowEventLevel["INFO"] = "INFO";
    FlowEventLevel["MILESTONE"] = "MILESTONE";
    FlowEventLevel["WARN"] = "WARN";
    FlowEventLevel["ERROR"] = "ERROR";
})(FlowEventLevel || (FlowEventLevel = {}));
export var FlowEventTiming;
(function (FlowEventTiming) {
    FlowEventTiming["CONTINUE"] = "CONTINUE";
    FlowEventTiming["START"] = "START";
    FlowEventTiming["END"] = "END";
    FlowEventTiming["CANCEL"] = "CANCEL";
    FlowEventTiming["IN_PROCESS"] = "IN_PROCESS";
})(FlowEventTiming || (FlowEventTiming = {}));
export var FlowEventType;
(function (FlowEventType) {
    FlowEventType["MONITORING"] = "MONITORING";
    FlowEventType["PROCESSING_REACTIONS"] = "PROCESSING_REACTIONS";
    FlowEventType["STARTING"] = "STARTING";
    FlowEventType["ONCE_ON"] = "ONCE_ON";
    FlowEventType["RUN_IF"] = "RUN_IF";
    FlowEventType["RUN"] = "RUN";
    FlowEventType["ADDING_REACTION"] = "ADDING_REACTION";
    FlowEventType["ERROR_USER_CODE"] = "ERROR_USER_CODE";
    FlowEventType["MONITOR_TRANSACTION"] = "MONITOR_TRANSACTION";
    FlowEventType["REQUESTING_STATUS"] = "REQUESTING_STATUS";
    FlowEventType["SETTLING_STATUS"] = "SETTLING_STATUS";
    FlowEventType["PROCESSING_STATUS"] = "PROCESSING_STATUS";
    FlowEventType["STATUS_REQUEST_COMPLETED"] = "STATUS_REQUEST_COMPLETED";
    FlowEventType["REQUESTING_TRANSITION"] = "REQUESTING_TRANSITION";
    FlowEventType["ROOT_REQUEST"] = "ROOT_REQUEST";
    FlowEventType["CREATING"] = "CREATING";
    FlowEventType["USER_REACTIONS"] = "USER_REACTIONS";
    FlowEventType["USER_CODE"] = "USER_CODE";
})(FlowEventType || (FlowEventType = {}));
